pkg/driver/node/regionprovider/provider.go

```diff
@@ -0,0 +1,69 @@
+// Package regionprovider provides utilities for detecting region by
+// looking environment variables, mount options, or calling IMDS.
+package regionprovider
+
+import (
+	"errors"
+
+	"k8s.io/klog/v2"
+
+	"github.com/awslabs/aws-s3-csi-driver/pkg/driver/node/envprovider"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/driver/node/volumecontext"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/mountpoint"
+)
+
+// ErrUnknownRegion is the error returned when the region could not be detected.
+var ErrUnknownRegion = errors.New("regionprovider: unknown region")
+
+// A Provider provides methods for detecting regions.
+type Provider struct {
+	regionFromIMDS func() (string, error)
+}
+
+// New creates a new [Provider] by using given [regionFromIMDS].
+func New(regionFromIMDS func() (string, error)) *Provider {
+	// `regionFromIMDS` is a `sync.OnceValues` and it only makes request to IMDS once,
+	// this call is basically here to pre-warm the cache of IMDS call.
+	go func() {
+		_, _ = regionFromIMDS()
+	}()
+
+	return &Provider{regionFromIMDS: regionFromIMDS}
+}
+
+// SecurityTokenService tries to detect AWS region to use for STS.
+//
+// It looks for the following (in-order):
+//  1. `stsRegion` passed via volume context
+//  2. Region set for S3 bucket via mount options
+//  3. `AWS_REGION` or `AWS_DEFAULT_REGION` env variables
+//  4. Calling IMDS to detect region
+//
+// It returns [ErrUnknownRegion] if all of them fails.
+func (p *Provider) SecurityTokenService(volumeContext map[string]string, args mountpoint.Args) (string, error) {
```

Comment:
The name here doesn't describe the functionality. From what I understand, it either returns the region to use *for* STS or returns an error. Perhaps `getSTSRegion`?

Reply:
I think `get` adds value as otherwise it's not obvious what we're actually doing with STS. I'd be happy with `regionProvider.getSTS(...)`